### PR TITLE
Add the bread and butter of the lingon linking layer

### PR DESCRIPTION
**Pull request by FredTheDino**

### CODE REVIEW

**Comment by sornas** [comment]
https://github.com/sornas/lingon/pull/12/commits/f8ffeeecdb4a378bda23218bee983428a22d4647 this might help.

File: sylt_macro/src/lib.rs [hunk 1]
```diff
@@ -329,3 +330,68 @@ pub fn derive_next(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
     };
     proc_macro::TokenStream::from(item)
 }
+
+use lazy_static::lazy_static;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
```
Comment (sornas) [inline sylt_macro/src/lib.rs]: Move these to the top.

File: sylt_macro/src/lib.rs [hunk 2]
```diff
@@ -329,3 +330,68 @@ pub fn derive_next(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+            }
+    }).collect();
+
+    let tokens = quote! {
+        pub fn _sylt_link() -> Vec<(String, RustFunction)> {
+            vec! [ #(#funs)* ]
+        }
+    };
+    eprintln!("RES\n {}", tokens.to_string());
+    proc_macro::TokenStream::from(tokens)
```
Comment (sornas) [inline sylt_macro/src/lib.rs]:
```suggestion
    proc_macro::TokenStream::from(tokens)
```

**Comment by sornas** [comment]
Wait, github showed me the older version. Let me take a look at it again.

File: sylt_macro/src/lib.rs [hunk 3]
```diff
@@ -1,8 +1,14 @@
 use std::path::Path;
+use std::str::FromStr;
 
 use quote::{format_ident, quote};
 use syn::{Expr, Pat, Token, parse::{Parse, ParseStream, Result}, parse_macro_input};
 
+use lazy_static::lazy_static;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
```
Comment (sornas) [inline sylt_macro/src/lib.rs:9]: I like to group my includes.

File: sylt_macro/src/lib.rs [hunk 4]
```diff
@@ -329,3 +343,108 @@ pub fn derive_next(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+        link
+    } else {
+        let tokens = quote! {
+            std::compile_error!("No functions to link, this call produces nothing.");
+        };
+        return proc_macro::TokenStream::from(tokens);
+    };
+    if matches!(link.state, LinkState::Written) {
+        let tokens = quote! {
+            std::compile_error!("Trying to write linked sylt functions twice, this is probably an error.");
```
Comment (sornas) [inline sylt_macro/src/lib.rs:386]:
```suggestion
            std::compile_error!("Tried to write linked sylt functions twice.");
```

Comment (sornas) [reply sylt_macro/src/lib.rs:386]: I feel like the error part is kind of given :)

File: sylt_macro/src/lib.rs [hunk 5]
```diff
@@ -329,3 +343,108 @@ pub fn derive_next(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+pub fn sylt_link_gen(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let module: syn::LitStr = parse_macro_input!(tokens);
+    let module = module.value();
+
+    let mut link = LINKS.lock().unwrap();
+    let mut link = if let Some(link) = link.get_mut(&module) {
+        link
+    } else {
+        let tokens = quote! {
+            std::compile_error!("No functions to link, this call produces nothing.");
```
Comment (sornas) [inline sylt_macro/src/lib.rs:380]:
```suggestion
            std::compile_error!("No functions to link. This call produces nothing.");
```

File: sylt_macro/src/lib.rs [hunk 6]
```diff
@@ -329,3 +343,108 @@ pub fn derive_next(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+pub fn sylt_link(attrib: proc_macro::TokenStream, tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let parsed: syn::ItemFn = parse_macro_input!(tokens);
+    let fun = parsed.sig.ident.clone();
+    let link: SyltLink = parse_macro_input!(attrib);
+
+    let mut links = LINKS.lock().unwrap();
+    let links = links.entry(link.module.value()).or_insert(ModuleLink::new());
+    if matches!(links.state, LinkState::Written) {
+        let tokens = quote! {
+            std::compile_error!("Trying to write linked sylt functions twice, this is probably an error.");
```
Comment (sornas) [inline sylt_macro/src/lib.rs:434]:
```suggestion
            std::compile_error!("Tried to write linked sylt functions twice.");
```

**Review by sornas — APPROVED** [review]

File: sylt_macro/src/lib.rs (hunk 3 above)
Comment (FredTheDino) [reply sylt_macro/src/lib.rs:9]: Instructions unclear...

**Pull request merged**